File: app/src/main/java/com/irishinterest/irishinterest/fragments/mainScreen/helper/endlessScroll/EndlessBookListAdapter.java
package com.irishinterest.irishinterest.fragments.mainScreen.helper.endlessScroll;

import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.paging.PagedListAdapter;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.irishinterest.irishinterest.fragments.mainScreen.BookDetailsActivity;
import com.irishinterest.irishinterest.R;
import com.irishinterest.irishinterest.model.books.Book;
import com.squareup.picasso.Picasso;

public class EndlessBookListAdapter extends PagedListAdapter<Book, EndlessBookListAdapter.ItemViewHolder> {

    private Context context;
    private Picasso picasso;

    public EndlessBookListAdapter(Context context){
        super(DIFF_CALLBACK);
        this.context = context;
        this.picasso = Picasso.with(context);
    }

    @NonNull
    @Override
    public EndlessBookListAdapter.ItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.book_list_item, parent, false);
        return new ItemViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull EndlessBookListAdapter.ItemViewHolder holder, int position) {
        Book book = getItem(position);
        if(book != null) {
            holder.bindView(book);
        }
    }

    private static DiffUtil.ItemCallback<Book> DIFF_CALLBACK =
            new DiffUtil.ItemCallback<Book>() {
                @Override
                public boolean areItemsTheSame(Book oldItem, Book newItem) {
                    return oldItem.getId() == newItem.getId();
                }

                @RequiresApi(api = Build.VERSION_CODES.KITKAT)
                @Override
                public boolean areContentsTheSame(Book oldItem, Book newItem) {
                    return oldItem.equals(newItem);
                }
            };

    class ItemViewHolder extends RecyclerView.ViewHolder {
        public ItemViewHolder(View v) {
            super(v);
        }

        private ImageView bookcover;
        private TextView title;
        private TextView author;

        public void bindView(Book book){
            title = itemView.findViewById(R.id.bookListItemTitleText);
            author = itemView.findViewById(R.id.bookListItemSubtitleText);
            bookcover = itemView.findViewById(R.id.bookListItemImageResource);

            picasso.load("https://irishinterest.ie/upload/" + book.getImage())
                    .placeholder(R.drawable.placeholder)
                    .into(bookcover);
           if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
               title.setText(Html.fromHtml(book.getTitle() != null ? book.getTitle().replace("'", "") : "", Html.FROM_HTML_MODE_COMPACT));
           } else {
               title.setText(Html.fromHtml(book.getTitle() != null ? book.getTitle().replace("'", "") : ""));
           }
            //title.setText(book.getTitle() != null ? book.getTitle().replace("'", "") : "");
            author.setText(book.getAuthor() != null && book.getAuthor().isEmpty() ? book.getPublisher() : book.getAuthor());

            final Book tmpBook = book;
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Intent intent = new Intent(context, BookDetailsActivity.class);
                    if(tmpBook.getTitle() != null) {
                        intent.putExtra("bookTitle", tmpBook.getTitle().replace("'", ""));
                    } else {
                        intent.putExtra("bookTitle", "");
                    }
                    intent.putExtra("bookAuthor", tmpBook.getAuthor().isEmpty() || tmpBook.getAuthor().equals(" ") ? tmpBook.getPublisher() : tmpBook.getAuthor());
                    intent.putExtra("bookPublisher", tmpBook.getPublisher());
                    intent.putExtra("bookPages", tmpBook.getPages());
                    intent.putExtra("bookPublishDate", tmpBook.getPublished());
                    intent.putExtra("bookAvailableAs", tmpBook.getPaperback() == 0 ? "Hardcover" : "Paperback");
                    intent.putExtra("bookCategory", tmpBook.getGenre());
                    intent.putExtra("bookDescription", tmpBook.getSynopsis());
                    intent.putExtra("bookImage", tmpBook.getImage());
                    intent.putExtra("bookAmazonPage", tmpBook.getVendorurl());
                    intent.putExtra("isbn13", tmpBook.getIsbn13());
                    intent.putExtra("bookId", tmpBook.getId());
                    intent.putExtra("authorId", tmpBook.getAuthorid());
                    intent.putExtra("bookIsbn", tmpBook.getIsbn());
                    intent.putExtra("bookUsAsin", tmpBook.getUsAsin());
                    intent.putExtra("bookUkAsin", tmpBook.getUsAsin());
                    intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                    context.startActivity(intent);
                }
            });
        }
    }
}

File: app/src/main/java/com/irishinterest/irishinterest/fragments/mainScreen/helper/endlessScroll/EndlessBookListAdapter.kt
package com.irishinterest.irishinterest.fragments.mainScreen.helper.endlessScroll

import android.content.Context
import androidx.paging.PagedListAdapter
import com.irishinterest.irishinterest.fragments.mainScreen.helper.endlessScroll.EndlessBookListAdapter.ItemViewHolder
import com.irishinterest.irishinterest.fragments.mainScreen.helper.endlessScroll.EndlessBookListAdapter
import com.squareup.picasso.Picasso
import android.view.ViewGroup
import android.view.LayoutInflater
import com.irishinterest.irishinterest.R
import androidx.recyclerview.widget.RecyclerView
import android.widget.TextView
import android.os.Build
import android.text.Html
import android.content.Intent
import android.view.View
import android.widget.ImageView
import com.irishinterest.irishinterest.fragments.mainScreen.BookDetailsActivity
import androidx.recyclerview.widget.DiffUtil
import androidx.annotation.RequiresApi
import com.irishinterest.irishinterest.model.books.Book

class EndlessBookListAdapter(private val context: Context) :
    PagedListAdapter<Book, ItemViewHolder>(
        DIFF_CALLBACK
    ) {
    private val picasso: Picasso
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemViewHolder {
        val view = LayoutInflater.from(context).inflate(R.layout.book_list_item, parent, false)
        return ItemViewHolder(view)
    }

    override fun onBindViewHolder(holder: ItemViewHolder, position: Int) {
        val book = getItem(position)
        if (book != null) {
            holder.bindView(book)
        }
    }

    inner class ItemViewHolder(v: View?) : RecyclerView.ViewHolder(
        v!!
    ) {
        fun bindView(book: Book) {
            val title: TextView = itemView.findViewById(R.id.bookListItemTitleText)
            val author: TextView = itemView.findViewById(R.id.bookListItemSubtitleText)
            val bookcover: ImageView = itemView.findViewById(R.id.bookListItemImageResource)
            picasso.load("https://irishinterest.ie/upload/" + book.image)
                .placeholder(R.drawable.placeholder)
                .into(bookcover)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                title.setText(
                    Html.fromHtml(
                        if (book.title != null) book.title!!.replace(
                            "'",
                            ""
                        ) else "", Html.FROM_HTML_MODE_COMPACT
                    )
                )
            } else {
                title.setText(
                    Html.fromHtml(
                        if (book.title != null) book.title!!.replace(
                            "'",
                            ""
                        ) else ""
                    )
                )
            }
            //title.setText(book.getTitle() != null ? book.getTitle().replace("'", "") : "");
            author.setText(if (book.author != null && book.author!!.isEmpty()) book.publisher else book.author)
            itemView.setOnClickListener {
                val intent = Intent(context, BookDetailsActivity::class.java)
                if (book.title != null) {
                    intent.putExtra("bookTitle", book.title!!.replace("'", ""))
                } else {
                    intent.putExtra("bookTitle", "")
                }
                intent.putExtra(
                    "bookAuthor",
                    if (book.author!!.isEmpty() || book.author == " ") book.publisher else book.author
                )
                intent.putExtra("bookPublisher", book.publisher)
                intent.putExtra("bookPages", book.pages)
                intent.putExtra("bookPublishDate", book.published)
                intent.putExtra(
                    "bookAvailableAs",
                    if (book.paperback == 0) "Hardcover" else "Paperback"
                )
                intent.putExtra("bookCategory", book.genre)
                intent.putExtra("bookDescription", book.synopsis)
                intent.putExtra("bookImage", book.image)
                intent.putExtra("bookAmazonPage", book.vendorurl)
                intent.putExtra("isbn13", book.isbn13)
                intent.putExtra("bookId", book.id)
                intent.putExtra("authorId", book.authorid)
                intent.putExtra("bookIsbn", book.isbn)
                intent.putExtra("bookUsAsin", book.usAsin)
                intent.putExtra("bookUkAsin", book.usAsin)
                intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                context.startActivity(intent)
            }
        }
    }

    companion object {
        private val DIFF_CALLBACK: DiffUtil.ItemCallback<Book> =
            object : DiffUtil.ItemCallback<Book>() {
                override fun areItemsTheSame(oldItem: Book, newItem: Book): Boolean {
                    return oldItem.id === newItem.id
                }

                @RequiresApi(api = Build.VERSION_CODES.KITKAT)
                override fun areContentsTheSame(oldItem: Book, newItem: Book): Boolean {
                    return oldItem.equals(newItem)
                }
            }
    }

    init {
        picasso = Picasso.get()
    }
}